pkg/kritis/admission/admission.go

```diff
@@ -18,45 +18,149 @@ package admission
 
 import (
 	"encoding/json"
-	"log"
-	"net/http"
-
+	"fmt"
 	"github.com/grafeas/kritis/pkg/kritis/admission/constants"
+	kritisv1beta1 "github.com/grafeas/kritis/pkg/kritis/apis/kritis/v1beta1"
+	"github.com/grafeas/kritis/pkg/kritis/crd/securitypolicy"
+	"github.com/grafeas/kritis/pkg/kritis/kubectl/plugins/resolve"
+	"github.com/grafeas/kritis/pkg/kritis/metadata"
+	"github.com/grafeas/kritis/pkg/kritis/metadata/containeranalysis"
+	"github.com/grafeas/kritis/pkg/kritis/pods"
+	"io/ioutil"
 	"k8s.io/api/admission/v1beta1"
+	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"log"
+	"net/http"
 )
 
-// This is very basic admission controller which allows all pods.
+type config struct {
+	retrievePod           func(r *http.Request) (*v1.Pod, error)
+	metadata              func() (metadata.MetadataFetcher, error)
+	imagesecuritypolicies func(namespace string) ([]kritisv1beta1.ImageSecurityPolicy, error)
+	validate              func(isp kritisv1beta1.ImageSecurityPolicy, project, image string, client metadata.MetadataFetcher) ([]metadata.Vulnerability, error)
+}
+
+var (
+	// For testing
+	admissionConfig = config{
+		retrievePod:           pod,
+		metadata:              metadataClient,
+		imagesecuritypolicies: securitypolicy.ImageSecurityPolicies,
+		validate:              securitypolicy.ValidateImageSecurityPolicy,
+	}
+)
+
+// This admission controller looks for the breakglass annotation
+// If one is not found, it validates against image security policies
+// TODO: Check for attestations
 func AdmissionReviewHandler(w http.ResponseWriter, r *http.Request) {
-	// TODO: Add image security policy check
-	// 1. Get all Image security Polices
-	// 2. Get all pod specs
-	// 3. For all pods Get Vulnerabilities for images
-	// 4. Validate policy
-	// 5. Attest.
-	status := &v1beta1.AdmissionResponse{
-		Allowed: true,
+	pod, err := admissionConfig.retrievePod(r)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	// First, check for a breakglass annotation on the pod
+	if breakglass(pod) {
+		returnStatus(constants.SuccessStatus, constants.SuccessMessage, w)
+		return
+	}
+	// Second, get all images in the pod and make sure they are fully qualified
+	images := pods.Images(*pod)
+	for _, image := range images {
+		if !resolve.FullyQualifiedImage(image) {
+			returnStatus(constants.FailureStatus, fmt.Sprintf("%s is not a fully qualified image", image), w)
+			return
+		}
+	}
+	// Third, validate images in the pod against ImageSecurityPolicies in the same namespace
+	isps, err := admissionConfig.imagesecuritypolicies(pod.Namespace)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
```

Comment:
Can we  log the error here?
You can simply do log.Printf()

Reply:
done